lib/data_update_scripts/20200910140109_cleanup_published_articles_with_duplicate_user_id_title_body_markdown.rb

```diff
@@ -0,0 +1,104 @@
+module DataUpdateScripts
+  class CleanupPublishedArticlesWithDuplicateUserIdTitleBodyMarkdown
+    def run
+      # load all published duplicates
+      rows = ActiveRecord::Base.connection.execute(
+        <<~SQL.squish,
+          SELECT *
+          FROM
+              (SELECT *,
+                      ROW_NUMBER() OVER(PARTITION BY user_id, title, body_markdown) AS row_number,
+                      COUNT(*) OVER(PARTITION BY user_id, title, body_markdown) AS num_rows_in_partition
+               FROM articles) duplicates
+          WHERE published = 't'
+              AND num_rows_in_partition > 1 -- we select all rows in each partition with duplicates
+        SQL
+      )
+
+      articles_to_delete_ids = []
+      ActiveRecord::Base.transaction do
+        # now that we have all rows in the partition, we group them by the partition columns
+        # so we can have small groups referring to the same article and graft data of surplus articles
+        # onto the first of the partition (it doesn't really matter which one we're going to use as the "mold")
+        groups = rows.group_by { |row| [row["user_id"], row["title"], row["body_markdown"]] }
+
+        # for each group we select the first, and then graft the data of the other ones on to it
+
+        groups.each_value do |group|
+          article_to_keep = group.first
+          articles_to_graft = group[1..]
+
+          article = Article.find(article_to_keep["id"])
+
+          graft_articles(article, articles_to_graft)
+
+          # save all the magic
+          article.save
+          article.index_to_elasticsearch_inline
+
+          articles_to_delete_ids += articles_to_graft.map { |a| a["id"] }
+        end
+
+        # we shouldn't need to call Articles::Destroy as all the data has been moved over by now
+        Article.where(id: articles_to_delete_ids).destroy_all
+      end
+
+      return if articles_to_delete_ids.blank?
+
+      # Store deleted IDs temporarily in Redis for safe keeping
+      Rails.cache.write(
+        "DataUpdateScripts::CleanupPublishedArticlesWithDuplicateUserIdTitleBodyMarkdown",
+        articles_to_delete_ids,
+        expires_in: 2.weeks,
+      )
+    end
+
+    private
+
+    # Main goal here is copy and merge everything from the duplicates to the main one
+    # NOTE: I'm intentionally ignore the case that two duplicates belong to different orgs
+    # or different collections as it's highly unlikely
+    def graft_articles(article_to_keep, articles_to_graft)
```

Comment:
This transaction block has a few issues. 
1) Duplicate Notification Subscriptions when we try to roll them over `duplicate key value violates unique constraint "idx_notification_subs_on_user_id_notifiable_type_notifiable_id"` 
2) Duplicate Rating votes when we try to roll them over `duplicate key value violates unique constraint "index_rating_votes_on_user_id_and_article_id_and_context"`
3) Duplicate Notifications when we try to roll them over `duplicate key value violates unique constraint "index_notifications_on_user_notifiable_and_action_not_null"` - Honestly, I think we don't roll these over and we let them be destroyed. 

I tried to fix the first couple by doing this within the transaction block. 
```ruby
  NotificationSubscription.where(
    notifiable_type: "Article",
    notifiable_id: articles_to_graft_ids
  ).each do |notification_sub|
    notification_sub.update(notifiable_id: article_id)
  rescue ActiveRecord::RecordNotUnique
  end
```
However, that does not work bc as it turns out if Postgres raises an exception, it poisons future transactions on the same connection and you end up with this error
```
ActiveRecord::StatementInvalid (PG::InFailedSqlTransaction: ERROR:  current transaction is aborted, commands ignored until end of transaction block)
```

TL;DR We need a way to avoid duplicates when we are rolling over records and my 6pm brain is not able to figure it out right now ;)

Reply:
> if Postgres raises an exception, it poisons future transactions on the same connection

This is a setting called `ON_ERROR_ROLLBACK`.

When it's off statements after the error will not be executed:

```
PracticalDeveloper_development# begin;
BEGIN
PracticalDeveloper_development*# select 1/0;
ERROR:  division by zero
PracticalDeveloper_development!# select 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
PracticalDeveloper_development!# rollback;
ROLLBACK
```

When it's on, statements continue executing (this implicitly uses [`ROLLBACK TO SAVEPOINT`](https://www.postgresql.org/docs/12/sql-rollback-to.html)):

PracticalDeveloper_development# \set ON_ERROR_ROLLBACK
PracticalDeveloper_development# begin;
BEGIN
PracticalDeveloper_development*# select 1/0;
ERROR:  division by zero
PracticalDeveloper_development*# select 1;
┌──────────┐
│ ?column? │
╞══════════╡
│        1 │
└──────────┘
(1 row)

PracticalDeveloper_development*# commit;
COMMIT
```

That said, option 3 seems sensible to me.

Reply:
Thanks a lot @mstruve for the test! So glad unique indexes in the DB are working, we wouldn't have never caught these duplicates if we'd only have the Rails layer :D All this work we've been doing is validated by these errors.

I feel we have 2 options here: 

1. use a similar technique as with tags (check if they already exists, if they don't add them, otherwise skip)
1. ignore these three models and forget about it

I'll set a time slot to see if I can come up with a simple logic, otherwise we move to plan B, though being able to merge those as well should be helpful if we decide to move this script behind an admin page in the end

Reply:
Decided to ignore NotificationSubscription, Notification and RatingVote :)